sprinkle/low-level/context.py

```diff
@@ -0,0 +1,127 @@
+"""
+CONTEXT OBJECTS IN THE SPRINKLE API
+
+The context object is core to how we reverse the plugin experience into a
+library experience; the context object is sort of like an instance of a
+library that has some information bound within it that is particular to the
+current running task.  E.g. det.get_trial_id() would be a strange library call
+but context.get_trial_id() is very natural.
+
+Goals for context objects in the Sprinkle API:
+
+ - There should only be one context class for all things PyTorchTrial, and
+   only one for all things Keras, etc.  The context class for PyTorchTrial and
+   the context class for Keras don't have to be the same though.
+
+ - The same context object must be usable in both local training and cluster
+   training environments.  Obviously if it exposes cluster-training-specific
+   extensions during cluster training, those extensions do not need to work in
+   local training environments, but any user code which does not depend on
+   such environment-specific extensions should run equally well in either
+   environment.
+
+ - The same context object must be usable in both training and non-training
+   workloads.  The sprinkle API for training should feel analagous to the
+   sprnkle API for distributed batch inference and to distributed test set
+   metrics, even though the calls the user makes will differ.
+
+ - Backwards compatibility: the existing Trial APIs need to still work, so
+   those TrialContext classes should remain unchanged.
+"""
+
+# GENERIC CONTEXT
+# Useful for custom workloads and is the basis for the framework contexts.
+context = det.GenericContext()
+
+# distributed data: applies to any workload; training or inference, local or cluster.
+# (same as the DistributedContext right now)
+context.distributed.get_rank()
+context.distributed.get_size()
+context.distributed.get_local_rank()
+context.distributed.get_num_agents()
+
+# rendezvous data: info about machines and IPs that are participating in distributed jobs.
+# This is configured automatically by the rendezvous info layer in Determined, but can also be
+# configured automatically through an environment variable if you want to use Determined's dtrain
+# features outside of Determined.
+context.rendezvous.get_num_nodes()
+context.rendezvous.get_node_rank()
+context.rendezvous.get_slots_per_node()
+
+# master data: applies to any job running on-cluster, of any job that you have
+# chosen to connect to the master with context.connect_to_master()
+context.connect_to_master(...)  # args similar to Determined() object
```

Comment:
(I mentioned this above as well) `connect_to_master()` is not going to land any time soon, but it is an answer to the question of "where would I feed my master url and creds to if I was training on my laptop but wanted to push metrics to a running master"

So accepting credentials in every call to `begin_training()` (before every reportable training workload) seems a bit strange to me.  Besides, `begin_training()` could potentially push to some local file backend when it's not pushing to the master, and that should be configured somewhere at the context level.

Hm.  Actually this would make a lot of sense to go into `init()`.